pydantic_ai_slim/pydantic_ai/ext/temporal/_agent.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations
+
+from collections.abc import Generator
+from contextlib import contextmanager
+from typing import Any, Callable
+
+from pydantic_ai.agent import Agent
+from pydantic_ai.models import Model
+from pydantic_ai.toolsets.abstract import AbstractToolset
+
+from ._model import temporalize_model, untemporalize_model
+from ._settings import TemporalSettings
+from ._toolset import temporalize_toolset, untemporalize_toolset
+
+
+def temporalize_agent(
+    agent: Agent[Any, Any],
+    settings: TemporalSettings | None = None,
+    toolset_settings: dict[str, TemporalSettings] = {},
+    tool_settings: dict[str, dict[str, TemporalSettings]] = {},
+    temporalize_toolset_func: Callable[
+        [AbstractToolset, TemporalSettings | None, dict[str, TemporalSettings]], list[Callable[..., Any]]
+    ] = temporalize_toolset,
+) -> list[Callable[..., Any]]:
+    """Temporalize an agent.
+
+    Args:
+        agent: The agent to temporalize.
+        settings: The temporal settings to use.
+        toolset_settings: The temporal settings to use for specific toolsets identified by ID.
+        tool_settings: The temporal settings to use for specific tools identified by toolset ID and tool name.
+        temporalize_toolset_func: The function to use to temporalize the toolsets.
+    """
+    if existing_activities := getattr(agent, '__temporal_activities', None):
+        return existing_activities
+
+    settings = settings or TemporalSettings()
+
+    activities: list[Callable[..., Any]] = []
+    if isinstance(agent.model, Model):
+        activities.extend(temporalize_model(agent.model, settings, agent._event_stream_handler))  # pyright: ignore[reportPrivateUsage]
```

Comment:
This will not work right if two agents using the same model are both temporalized -- only the event stream handler for the first call will stick :grimace: